File: voc-mesh-platform/apps/api/app/models/billing.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class BillingEvent(Base):
    __tablename__ = "billing_events"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False, index=True
    )
    period_start: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    period_end: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    sensor_count: Mapped[int] = mapped_column(Integer, default=0)
    km2_covered: Mapped[float] = mapped_column(Float, default=0.0)
    amount_usd: Mapped[float] = mapped_column(Float, default=0.0)
    tier: Mapped[str] = mapped_column(String(64), nullable=False)
    paid: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
    )

File: voc-mesh-platform/packages/voc-schema/src/billing.ts
import { z } from "zod";
import { TenantTier } from "./tenant";

export const BillingEventSchema = z.object({
  id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  period_start: z.string().datetime(),
  period_end: z.string().datetime(),
  sensor_count: z.number(),
  km2_covered: z.number(),
  amount_usd: z.number(),
  tier: TenantTier,
  paid: z.boolean(),
});
export type BillingEvent = z.infer<typeof BillingEventSchema>;
